### ISSUE TITLE
[Umbrella Issue] Potential Code Improvements

### ISSUE THREAD

**Issue by MadhavJivrajani** [dims/maintainers#12]
While playing around with the tool and going through the code, there are a few areas we can try and polish further:

### Code Quality

- [Nit] Readme says "`--dry-run=true` to update files", should be `--dry-run=false`
- If the tool is run on a K8s clone that has an `_output` folder, `panic`s occur. We should probably skip processing the `_output` folder 
- Speaking of `panic`s, to move away `panic(err)`, we can maybe look at using `RunE` instead of `Run` in the cobra [Command](https://pkg.go.dev/github.com/spf13/cobra#Command) struct
- We should consider introducing struct types that group flags per command instead of having multiple global variables defined (similar to how `kubectl` does it)
  - Speaking of flags, one instance of flag validation that I came across was in the `prune`  command, both `--skip-github` and `--skip-devstats` should not be allowed to be set together maybe, validations like this can be done in a cleaner way with struct types and corresponding receivers

### Code Testability

Eventually, we will need to make sure that there is good test coverage for the tooling to safely be able to add more functionality. IMO we'll need to define interfaces for functionality and mock them for testing. A similar refactor was done here:
- https://github.com/kubernetes/k8s.io/pull/2713
- https://github.com/kubernetes/k8s.io/pull/2873

### Up For Discussion

I've noticed that in a few places like
https://github.com/dims/maintainers/blob/78c7c69225db08d6388138c4d8df3fd5006acc0c/pkg/utils/github_utils.go#L79
a pointer to a slice is returned, is there a reason for returning a pointer to a slice?

And another instance of passing a pointer to a map
https://github.com/dims/maintainers/blob/78c7c69225db08d6388138c4d8df3fd5006acc0c/cmd/validate.go#L121

**Comment by nikhita** [comment]
+1 overall.

>  a pointer to a slice is returned, is there a reason for returning a pointer to a slice?

Agree we can remove the pointer. cc @dims in case you had specific reasons for this.

**Comment by nikhita** [comment]
If anyone wants to pick up this issue, please pick **one or two** items from the list above at a time. Please do _not_ send a PR with fixes for all of these issues at once.

**Comment by dims** [comment]
@nikhita @MadhavJivrajani +1 to cleanup pointer to a slice!

**Comment by RaghavRoy145** [comment]
One of the things mentioned in the description is functions that are returning a pointer to a slice can just return the slice itself, and I think this is because slices are pass by reference, but whenever I've used slices I was always asked to return pointers to slices as this ensures that even if the metadata of the slice itself is modified (like if the slice is reallocated after the slice is modified in a separate function), simply returning a slice would lead to unexpected behavior
This is nicely elaborated in [this medium article](https://medium.com/swlh/golang-tips-why-pointers-to-slices-are-useful-and-how-ignoring-them-can-lead-to-tricky-bugs-cac90f72e77b), I wanted to know if there's anything I'm missing here before I create a PR for this change.

This discussion was also taken forward in [this](https://kubernetes.slack.com/archives/C1TU9EB9S/p1645015323587279) slack thread and there were some interesting clarifications from @MadhavJivrajani

**Comment by dims** [comment]
a bit more context ... I don't remember exactly how i got there for some functions as i typically avoid pointers to slice. But yes, if we can get rid of it and still make things work. +1 from me. As i had to have added it explicitly, may be worth looking deeper and may be add unit tests as well.

**Comment by RaghavRoy145** [comment]
I will be creating a PR for using RunE, and a separate PR to fix the panics that occur when k8s has the _output folder.

Will also be looking deeper into functions returning slice pointers (maybe try adding unit tests as mentioned by @dims ), after that. But please go ahead and open any PRs for slice pointers as I'm not working on it right now.

**Comment by RaghavRoy145** [comment]
I needed some help reproducing one of the issues, which is the panics that occur when the tool is used on a cloned k8s repo when it has a  `_output` folder.
I ran `make kubectl `on the root of the repo to create the `_output` folder, but all the commands I've run using this tool run successfully, is there a particular one I can run?

**Comment by RaghavRoy145** [comment]
I was interested in increasing the test coverage for the tool,
Since I don't have a lot of experience in this area I wanted to know if the way to start creating tests for a tool is to create a fake client (like implemented here: https://github.com/kubernetes/k8s.io/pull/2873) that can run the tool with its different functionalities on fake files (or maybe real ones?).
- If yes, then to what extent do we create scenarios where the tool might fail?
- What kinds of tests can be added for a tool like this? Is a simple run of the different commands/flags enough?
- Are unit tests important? If yes then are there any good resources/PRs I can look at for reference?

**Comment by Atharva-Shinde** [comment]
Is this one of the [implementations](https://github.com/kubernetes/kubernetes/blob/7588a82bdb1f761ef2a591d18478b0145915af5d/staging/src/k8s.io/kubectl/pkg/cmd/create/create_role.go#L131) which is referenced here about kubectl of declaring structs rather than separate global variables @MadhavJivrajani ?

**Comment by MadhavJivrajani** [comment]
yes, something like that

**Comment by MadhavJivrajani** [comment]
@RaghavRoy145 
> If yes, then to what extent do we create scenarios where the tool might fail?

Its more about testing correctness, typically you write test cases that test out the functionality of individual components (typically individual functions). Its good to start with trying to cover the cases that a particular piece of code might output (its fine if it isn't 100% exhaustive yet, we can iterate on that)

> What kinds of tests can be added for a tool like this? Is a simple run of the different commands/flags enough?

No. Each command is a combination of multiple sub-functionalities. For example the `export` command is a combination of the `exportOwnersAndAliases` functionality. So we need to test _each_ individual functionality and then test the entire command on a high level as well.

> Are unit tests important? If yes then are there any good resources/PRs I can look at for reference?

Very important. The PR you linked adds unit tests itself. Look at the `_test.go` files and the corresponding code that it writes tests for (the test will call the fake client, but the functionality that it tries to mimic is what the real client implements https://github.com/kubernetes/k8s.io/blob/main/groups/service.go, so you try and look at the real client code to write tests for it but you use the fake client).
When should you want to use a fake client? When you deal with API calls because we can't make API calls to github as part of the tests, so in that case, we try to imitate or "mock" the behaviour of such calls. This is why we use interfaces here (the refactor linked in the PR description)

**Comment by RaghavRoy145** [comment]
This helps a lot, thanks! :smile:

**Comment by AvineshTripathi** [comment]
Hey @MadhavJivrajani for this "If the tool is run on a K8s clone that has an _output folder, panics occur. We should probably skip processing the _output folder " which command did you use and was it for k8s/k8s

**Comment by MadhavJivrajani** [comment]
Yes for k/k
```
❯ maintainers prune --skip-github --skip-devstats
Running script : 02-25-2022 17:40:40
panic: error processing /home/madhav/gocode/src/github.com/MadhavJivrajani/kubernetes/_output/local/go/pkg/mod/k8s.io/kube-openapi@v0.0.0-20211115234752-e816edb12b65/OWNERS: error unmarshaling JSON: while decoding JSON: json: unknown field "emritus_approvers"

goroutine 1 [running]:
github.com/dims/maintainers/cmd.glob..func6(0xab6260, {0x7d722c, 0x2, 0x2})
	github.com/dims/maintainers/cmd/prune.go:66 +0x1087
github.com/spf13/cobra.(*Command).execute(0xab6260, {0xc00005c360, 0x2, 0x2})
	github.com/spf13/cobra@v1.3.0/command.go:860 +0x5f8
github.com/spf13/cobra.(*Command).ExecuteC(0xab64e0)
	github.com/spf13/cobra@v1.3.0/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.3.0/command.go:902
github.com/dims/maintainers/cmd.Execute()
	github.com/dims/maintainers/cmd/root.go:32 +0x25
main.main()
	github.com/dims/maintainers/main.go:22 +0x17
```
I don't have an exact reproducer for you. Maybe try running `make update` in k/k and then try?
Also note that the panic won't happen once https://github.com/dims/maintainers/pull/13 is merged, but we still don't want to process the `_output` folder.

**Comment by AvineshTripathi** [comment]
Hey @MadhavJivrajani I did a bit exploration in kubectl about the struct and was confused if we are going to change the the code of prune.go to a structure similar to kubectl where they have a struct of flags and a constructor initialize the vlaue and so on or we are just willing to convert the global variables into a single struct and rest remains the same. If I missed anything let me know

**Comment by RaghavRoy145** [comment]
When we add tests for the commands, along with testing sub-functionalities like `warnFileMismatchesBetweenKubernetesRepoAndSigsYaml` in the `validate` command here:
https://github.com/kubernetes-purgatory/maintainers/blob/63773f8a6fd26b47df20611095cd4bf69d1960f3/cmd/validate.go#L73

Should we also add tests for `utils` like `GetSigsYamlFile` here:
https://github.com/kubernetes-purgatory/maintainers/blob/63773f8a6fd26b47df20611095cd4bf69d1960f3/cmd/validate.go#L51

And other functionalities like `PrefixToGroupMap` here:
https://github.com/kubernetes-purgatory/maintainers/blob/63773f8a6fd26b47df20611095cd4bf69d1960f3/cmd/validate.go#L71

Or is it enough to tests the main sub-functionalities defined in this file?

**Comment by MadhavJivrajani** [comment]
@AvineshTripathi 
> Hey @MadhavJivrajani I did a bit exploration in kubectl about the struct and was confused if we are going to change the the code of prune.go to a structure similar to kubectl where they have a struct of flags and a constructor initialize the vlaue and so on or we are just willing to convert the global variables into a single struct and rest remains the same. If I missed anything let me know

I haven't thought too much about refactoring things in a similar way to kubectl as of now. What I had in mind was 
> convert the global variables into a single struct

**Comment by AvineshTripathi** [comment]
Sounds good to me I ll do the change for the struct then

**Comment by RaghavRoy145** [comment]
While trying to write tests for some functions in the utils files, I noticed that the outputs of some functions, like https://github.com/kubernetes-sigs/maintainers/blob/dbe97f741b9bcfa148d32b70494efea98819cea2/pkg/utils/devstats_utils.go#L30
and https://github.com/kubernetes-sigs/maintainers/blob/dbe97f741b9bcfa148d32b70494efea98819cea2/pkg/utils/github_utils.go#L31
can keep changing as the number of comments/contributions change with time (in this example).

I wanted to know what's the best way to test the output of these kinds of functions, I thought of creating a fake repository with a fixed number of comments and contributions that these functions can query but I'm not sure if that's the best solution.

Another way could be to set up a test function to test against, that does the same thing as the function being tested should do. This will atleast take care of regressions, but that didn't sound very elegant. 
Any input regarding better ways to write tests for these kind of functions would be greatly appreciated!

**Comment by SD-13** [comment]
Hi @MadhavJivrajani, a newbie here. Do you have anything left here so I can give it a try? Thanks,

**Comment by Okabe-Junya** [comment]
Hi folks,
While checking the annual cleanup, I happened to run this tool and encountered another bug (reported in #31 )
I agree with almost all the suggestions in this issue for improving the code. I'll create PR for some of the easier fixes.
